sonar-plugin-api/src/main/java/org/sonar/api/server/rule/internal/DefaultNewRule.java

```diff
@@ -134,13 +136,48 @@ public DefaultNewRule setType(RuleType t) {
 
   @Override
   public RulesDefinition.NewRule addDescriptionSection(RuleDescriptionSection ruleDescriptionSection) {
-    checkArgument(isSectionKeyUnique(ruleDescriptionSection.getKey()), "A section with key %s already exists", ruleDescriptionSection.getKey());
+    assertRuleDescriptionSectionIsValid(ruleDescriptionSection);
     ruleDescriptionSections.add(ruleDescriptionSection);
     return this;
   }
 
+  private void assertRuleDescriptionSectionIsValid(RuleDescriptionSection ruleDescriptionSection) {
```

Comment:
This class is getting big and another idea would be to move these validation methods to another class i.e. DefaultNewRuleValidator or RuleDescriptionSectionsValidator.

On the other hand similar checks for other methods are already included in this class, so maybe it still counts as "single responbility".

Reply:
I don't have a strong opinion here. This class is indeed long but it's because it has a lot of public methods to set the different fields, not because it encapsulates complex logic. We could still externalize the validation. However, I'm not sure it will be more readable as we will have to supply the ruleDescriptionSections list as an extra argument.

Also, I indeed think the responsibility of this class is ok: it is responsible for the encapsulation of some data together with the constraints for those data. But I agree, the validation part could be split. @leo-geoffroy-sonarsource what's your opinion?

Reply:
No strong opinion either, another perspective is that as this is a public class quite central to the public API, maybe it is better to keep the meaningful information (fields, publics method) visible by externalizing the validators as it was done in PreConditions

Reply:
this class extends `RuleDefinition.NewRule` which defines the contract without noise... but I'm not sure about what you meant with `Preconditions` as it is already used by `DefaultNewRule`?

Reply:
org.sonar.api.utils.Preconditions is used in DefaultNewRule and aggregate a certain number of assert facilitators

Reply:
Suggestion from my side: https://github.com/SonarSource/sonar-plugin-api/pull/25/commits/ffe2bfadd84b318a36badedb619214b7964a730e (my apologies @aurelien-poscia-sonarsource for pushing it to this branch, we can squash later)

Reply:
Answer to Leo: `org.sonar.api.utils.Preconditions` is a generic static helper, it is not bound to any business logic. Here the need is a business-specific validation (which by the way would internally use `org.sonar.api.utils.Preconditions`). To be agreed if it should be embedded in `DefaultNewRule` or in a validator...

Reply:
The 3 of us agreed to leave it as it is. We will see if the other reviewers have concern about this piece of code.